pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go

```diff
@@ -83,10 +89,43 @@ func (s *preFilterState) Clone() framework.StateData {
 // failed by this plugin schedulable.
 func (pl *NodeAffinity) EventsToRegister() []framework.ClusterEventWithHint {
 	return []framework.ClusterEventWithHint{
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterNodeChange},
 	}
 }
 
+// isSchedulableAfterNodeChange is invoked whenever a node changed. It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *NodeAffinity) isSchedulableAfterNodeChange(pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	originalNode, modifiedNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		pl.logger.Error(err, "unexpected objects in isSchedulableAfterNodeChange", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	requiredNodeAffinity := nodeaffinity.GetRequiredNodeAffinity(pod)
+	wasMatched := false
+	if originalNode != nil {
+		wasMatched, err = requiredNodeAffinity.Match(originalNode)
+		if err != nil {
+			pl.logger.Error(err, "failed to match pod to the original node", "pod", klog.KObj(pod), "node", klog.KObj(originalNode))
+			return framework.QueueAfterBackoff
+		}
+	}
```

Comment:
Hi @sanposhiho, after I review the code, I'm confused the reason why we need check whether the old node is matched with the target pod. I think it's enough to check the new or updated node.

Reply:
If we only checked whether the updated/new node matches with the nodeaffinity or not, it would always return `QueueAfterBackoff` for any changes for nodes that match with the nodeaffinity.
So, we need to check the old node as well so that we only return `QueueAfterBackoff` for a event that makes the node matches with the nodeaffinity

Reply:
Maybe we should be loose on this because what we verified here is not all the possibilities, then it will happen like we used to pass the `requiredNodeAffinity.Match(node)`, now it still passed but actually it was break by `NodeAffinity.addedNodeSelector`. So check the newObj is enough I think.

Also considering rescheduling is more expensive than hint func, can we also include the logics about `addedNodeSelector`?

Reply:
I agree that include the logics about addedNodeSelector (actually I completely forgot about that, good catch.)
But, I disagree that only checking newObj. In the cluster, the Node's status is updated very frequently. (like due to the heartbeat)
If we only checked the newObj only, it'd cause Pods, rejected by NodeAffinity, to be moved out from the unschedQ every time the Node, which matches the NodeAffinity, gets updated.

Reply:
Fair, what I concern is we may miss some possibilities, like we return unschedulable in PreFilter, so at the least, we should not return skip in this situation. I may take it as below, but it's still under discussion below.
```
if !wasMatched && isMatched {
    return immediately
}

# Else
return queueAfterBackoff
```

FYI, IIRC we'll use node lease for heartbeat right now.

Reply:
> Maybe we should be loose on this because what we verified here is not all the possibilities, then it will happen like we used to pass the `requiredNodeAffinity.Match(node)`, now it still passed but actually it was break by `NodeAffinity.addedNodeSelector`. So check the newObj is enough I think.
> 
> Also considering rescheduling is more expensive than hint func, can we also include the logics about `addedNodeSelector`?

@sanposhiho @kerthcet added in https://github.com/kubernetes/kubernetes/compare/9acc52b7511bb9c475da0ca888e0955e3e69f7d1..fd4ce8ad0634f9d76fa4b7412ff2bdded9f8880a

Reply:
> we may miss some possibilities

I may not quite understand what @kerthcet meant, but maybe it's a similar topic to https://github.com/kubernetes/kubernetes/pull/119396#discussion_r1324024982?
As a conclusion of that thread, QueueingHints are supposed to return QueueAfterBackoff only when some transition happens, which makes Pod schedulable.

So, let's say NodeUpdate event happens and it changes NodeA from unmatch to match. Then we return QueueAfterBacckoff. But, say NodeUpdate event happens and it changes NodeA from match to match, we don't return QueueAfterBackoff, we return QueueSkip. 
If we heavily consider some unknown bugs that we miss such events like the first example, we should return QueueAfterBackoff for the second example though, we don't want do that -- as I said in this thread as well, that's too conservative and it'd result in putting back Pods to activeQ/backoffQ unnecessary via unrelated events very frequently.

---

Also talking about the difference between QueueAfterBackoff and QueueImmediately, there is one on-going semantic change to QueueingHint, which will simplify the returning value to have only Queue and QueueSkip.
https://github.com/kubernetes/kubernetes/pull/119517

For now until that PR gets merged, simply speaking, you should return QueueAfterBackoff in almost all cases in in-tree plugins, and return QueueImmediately only in DRA plugin. (You can see the conclusion of it in the above PR as well.)